### DISCUSSION TITLE
ARC-0042: Adjust block reward algorithm to make sure inflation is stable

### DISCUSSION THREAD

**Discussion by HarukaMa** [ProvableHQ/ARCs#76] [ARCs]
---
arc: 0042
title: Adjust block reward algorithm to make sure inflation is stable
authors: Haruka Ma
discussion: self
topic: Protocol 
status: Draft
created: 2024-09-09
---

## Abstract

The AleoBFT consensus doesn't guarantee a stable block interval, but the current block reward algorithm assumes the block interval is fixed at 10 seconds, and is emitting rewards regardless of actual block intervals. This makes the 5% annual reward target almost unattainable: shorter block intervals causes higher reward frequency and higher annual reward, and vice versa.

As an example, if the block interval is 3 seconds throughout a full year, the effective inflation rate from base block reward would be 16.67% instead of 5%.

This ARC proposes changes to the block reward algorithm so it could adjust the per-block reward according to the actual block interval, therefore achieving the 5% annual target without being affected by fluctuations of the validator network.

## Specification

Proposed new block reward algorithm and implementation:

```rs
/// Calculate the block reward, given the total supply, block interval, coinbase reward, and transaction fees.
///     R_staking = floor((0.05 * S) * min(I, 60) / S_Y) + CR / 3 + TX_F.
///     S = Total supply.
///     I = Seconds since last block.
///     S_Y = Seconds in a year (31536000).
///     CR = Coinbase reward.
///     TX_F = Transaction fees.
pub fn block_reward(
    total_supply: u64,
    secs_since_last_block: i64,
    coinbase_reward: u64,
    transaction_fees: u64,
) -> u64 {
    // Compute the annual reward: (0.05 * S).
    let annual_reward = total_supply / 20;
    // Compute the block reward: (0.05 * S) * min(I, 60) / S_Y.
    let block_reward = annual_reward * secs_since_last_block.min(60) as u64 / 31536000;
    // Return the sum of the block reward, coinbase reward, and transaction fees.
    block_reward + (coinbase_reward / 3) + transaction_fees
}
```

The proposed algorithm takes the actual block interval into account, emitting less reward if block interval is short, and vice versa. 

Note the block interval is capped to 60 seconds in the calculation; this is set mainly to remedy situations like chain halts or the gap between genesis timestamp and first block (only applicable to future chain resets for testnets) so there won't be a sudden large sum of credits being rewarded in one block. The 60 seconds here is just an arbitrary number; actual interval cap should be discussed further.

Also note the block interval is guaranteed to be greater than zero on the consensus level.

### Test Cases

Tested with the above code on local devnet:
![image](https://github.com/user-attachments/assets/6b0ffd7e-9d89-4de3-bf4a-b091e4c3f4ee)

## Dependencies

snarkVM. Potentially other products that relies on the implementation.

### Backwards Compatibility

Implementing this ARC requires the snarkVM and snarkOS to have a proper soft fork mechanism.

## Security & Compliance

The proposed algorithm affects the tokenomics, albeit it's trying to bring the chain back to the intended state.

### Conflict of interests among validators

This section lists some situations the author could think of.

1. As the block reward linearly increases with the block interval, validators might want to hold the block generation until the interval cap is reached. The only incentive here could be to save some CPU cycles related to block generation, as in the end the total block rewards will be the same regardless of the blocks generated in a given period. Plus, holding back the blocks might cause additional puzzle solutions to abort, as there could be at most 4 solutions in a block, and considering puzzle rewards contributes to the block reward as well, doing this could actually harm the income of validators.
2. Validators might collude to generate blocks with fake timestamps in order to inflate the rewards. In short-term, this might give validators more rewards. However, by being dishonest, validators will be harming the credibility of the chain; therefore in the long run, this would harm the investment done by validators considering the entry barrier of becoming one. Setting the interval cap to the previously intended block interval (10 seconds) might deter this as well.

## References

N/A

**Comment by Fatman13** [comment]
> ... the effective inflation rate from base block reward would be 16.67% instead of 5%.

Is the "base block reward" here referring to coinbase rewards in the graph at the bottom [here](https://aleo.org/aleo-credits/)? Does it mean that coinbase reward targets 5% and staking rewards targets 7% thus a combined 12% inflation rate?

**Reply by HarukaMa** [reply]
Stake reward should be fixed 5% per year, where coinbase reward is from puzzles (a third goes to block reward, two thirds go to puzzle reward)

**Comment by mwyatt896** [comment]
Put this in discord and also putting here:

some interesting numbers for you from our end over at Puzzle. 
TLDR;
looks like for every 10k we add to our Arcade WAU numbers, we add 50k txns and increase consensus block times by ~0.5s. Because of this fact -- in our opinion -- this ARC should be held off until November when we have more data from Coinbase Quest full launch and it's effects on consensus block times to see if they are in line with what we see.

Background on data below:
October 2nd -- we thought most mainnet kinks were worked out so we kicked off Puzzle Arcade campaigns with communities we've been working with for past few months
Oct 2 & Oct 3rd our costs blew up from paymaster and oct 4th we had to ask communities to stop using/chill out -- was costing us too much and validator rewards wouldn't have chance of covering at current delegation
see here: https://x.com/lets_puzzle/status/1842360983605637450

so in light of adjusting block reward algorithm we ran numbers to see how this increase in activity effected consensus and run quick analysis on projections of what it will do to consensus times/block rewards down road

October 2nd:
25,568 Puzzle txns
3.04 s/block
start block: https://aleoscan.io/block?h=798610
end block: https://aleoscan.io/block?h=827030

October 3rd: ~11k Arcade WAUs
50,726 Puzzle txns
3.41 s/block
start block: https://aleoscan.io/block?h=827030
end block: https://aleoscan.io/block?h=852349

Back of napkin projections:
20k Arcade WAUs
100k Puzzle txns
~4 s/block

30k Arcade WAUs
150k Puzzle Txns
~4.5 s/block

40k Arcade WAUs
200k Puzzle Txns
~5 s/block

Based on this quick analysis -- we think it's best to wait and see how this is effected with the coinbase ramp with Leo Wallet, Pondo, Verulink, and Arcane during this month.

If their numbers lineup with ours and based on number of transactions we expect coinbase to see alone, we are likely to hit 5s/block in November. And that is not counting what we are trying to grow to in November at Puzzle as well. There is a chance we get to 10s/block based on activity alone -- not even counting increasing validator set.

That being said about the near term, I do like this as a direction to continue to pursue as I imagine pushing validator sets + increasing usage might have us greater than 10s/block given our own internal targets lol. I believe this could be a good solution for both near and long term.

I do worry about whether this block reward calculation would work given that secs_since_last_block might be different between each node though. 

Any idea on if this would be an issue or any solutions to this potential issue?

**Comment by vicsn** [comment]
Big +1 from me for fixing the inflation issue.

~I'm afraid block reward can't easily be made a function of blocktime.~

~The validator rewards is something which all validators need to have equal view and opinion on. Right now that's trivially the case, because its a constant. But if reward becomes a function of blocktime, then how are we sure all validators observe the *exact* same blocktime? There is no such thing as a single canonical block timestamp. One thing which might be possible is to use the BatchCertificate time instead, which per definition have been signed by a quorum of validators. They contain a timestamp, so in theory the reward could be adjusted based on the speed of certificate generation. But it would still be only a very imperfect proxy of actual blocktimes. Making rewards depending on any condition introduces a forking risk and requires lots of resources to vet its security.~

~In contrast, making a node's local minimum batch delay a function of its local view of either batch/certificate/block does not introduce forking risk.~

UPDATE: actually maybe the timestamp is deterministic: https://github.com/AleoNet/snarkVM/tree/mainnet/ledger/narwhal/subdag/src/lib.rs#L191

**Reply by HarukaMa** [reply]
Why the block timestamp in the header can't be considered "canonical"? AFAIU it should be derived from batches that all validators agreed with, thus should be deterministic?

**Comment by apruden2008** [comment]
While I support the approach outlined in ARC-0042 for dynamic blocktime adjustments as a medium-term solution, I think we need an immediate band-aid to address the current issue of skewed reward distribution. By increasing blocktimes to around 8 seconds (buffered to 10 seconds under heavy load), we can better align rewards with Aleo’s original tokenomics model.

The implementation can be found here: https://github.com/AleoNet/snarkOS/pull/3412

This interim fix ensures rewards are balanced more evenly between validators and provers, while allowing time to implement the more comprehensive dynamic solution proposed above.

Therefore I propose merging PR #3412 and test in Canary Net, testnet beta according to the same test plan we ran for our mainnet roadmap.

**Reply by WietzeSlagman** [reply]
Getting a quick fix in to allign the economics better asap should be a good plan, getting close to the publicly shared tokenomics model should be a high priority.

However, this change and increasing the validator could lead to an undesirably high block time (more than 10s) and is something we should test for thoroughly on Canary and Testnet Beta. Potentially leading to pushing the increased validator set to after/with this ARC implementation go live or reverting/updating this change.

**Comment by dylanschultzie** [comment]
I generally support this proposal. 

One outstanding question I have, following Matt's comment, is what effect adding the 9 additional validators will have on block time. This isn't a "blocking" question, but it does make me wonder how pertinent this ARC will be in the coming months.

Does it make sense to add the 9 new validators to canarynet to get a feel for how that'll change the network speed?

**Reply by HarukaMa** [reply]
This ARC aims to decouple rewards from block time, and should work with any network condition (if the block interval is less than 60s).

**Reply by apruden2008** [reply]
Yes, I think that it would make the most sense to add the 9 validators AND push this change at the same time.

**Comment by Meshiest** [comment]
I was initially skeptical about the reliance on timestamps as they are notoriously untrustworthy, but the weighted median timestamp calculation from the certificates should mitigate validator collusion attacks. I would trust timestamps more if the delta times between timestamps were guaranteed to be statistically consistent across validators, but for now this _looks_ like a decent approach.

**Comment by OzielLa** [comment]
I generally support this proposal. 
We need to consider the calculation of coinbase_reward, where anchor_block_reward_at_height is also related to BLOCK_TIME and ANCHOR_TIME. Given the current block generation speed of around 3 seconds, by the end of the third year, we will reach a height of 28382400. The coinbase_reward is decreasing too rapidly.

**Reply by HarukaMa** [reply]
Yep, that needs to be addressed as well, but not sure if it should be included in this ARC right now

**Comment by howardwu** [comment]
I support @HarukaMa's proposal for a dynamic adjustment to the block reward.

As this ARC is intended to re-align the network's rewards with the published inflation schedule, I agree with @OzielLa's point that the `BLOCK_TIME` needs to be modified from `10` to `3` seconds. As this proposal will anchor the current block time at 3 seconds (rather than 10 seconds), making the fix to `BLOCK_TIME` will ensure the puzzle reward schedule does not de-accelerate at 3x the expected rate. **By combining these modifications, this achieves the intended goal of adhering to the published inflation schedule.**

@HarukaMa's points about potential collusion risks are correct. From my understanding, if a subset of validators collude, the network runs into the risk that forking is a possible concern. If at least 2/3rds of validators (by stake weight) collude, then the network runs into both a forking and a deviation from the anticipated inflation rate. Before rolling in these changes, adequate testing of these edge cases should be done to understand the complexity of pulling off such an attack, which in my opinion, also introduces more fundamental concerns to the sanctity of the network (CC @vicsn).

This change will require validators to update to a new version of snarkOS/VM with this ARC applied. The good news is that it does not require a coordinated synchronous update. Validators should be able to asynchronously update in advance of the target block height with which the rule change would occur. At least 2/3rds of all validators (by stake weight) will need to update to this version for the change to be effective on the network.

**Reply by HarukaMa** [reply]
For coinbase rewards, should we change the logic to use absolute timestamp as well (we can have both genesis timestamp and current block timestamp to calculate the intended coinbase reward at that time)? AFAIU the average block time might organically increase as we add more validators and with more transactions, and it seems we are going to have a short-term mitigation of artificial block delay before this ARC is implemented.

**Reply by zkxuerb** [reply]
The aritificial block delay is not going to be added in @HarukaMa.

**Reply by howardwu** [reply]
@HarukaMa one way to navigate around the puzzle reward timestamp concern is to use a locally-defined range. So instead of looking at the average block time from genesis to tip, one can choose to look purely at the prior block up to tip (so look back 1 block only). IIRC, this may be how Bitcoin Cash does their target readjustment algorithm.

In either case, an expected `BLOCK_TIME` variable is required to be baked into the source code, which I think is the bigger constraint in the design space here. Because unlike PoW networks, there is no real concept of an expected block time in a Bullshark & Narwhal blockDAG protocol.

**Reply by HarukaMa** [reply]
@howardwu considering the coinbase reward itself is supposed to be a linear function of time, given two timestamps, we can directly get the expected reward value without having to calculate average block times and such. I'm not sure why we still need the BLOCK_TIME for the coinbase reward?

**Reply by howardwu** [reply]
The reason is that the protocol needs a reference point (i.e. `BLOCK_TIME`) to know whether the network is moving faster or slower than expected. If the network is moving faster, then the difficulty needs to be adjusted up. If the network is moving slower, then the difficulty needs to be adjusted down.

**Reply by LoiusZK** [reply]
According to my understanding and what @OzielLa said, if `block_time` will be adjusted to 3 seconds then wouldn't the effect be the same as it is now? It does not reduce the current excessive POS output. At the same time, it will cause the coinbase reward to drop too quickly, which is not in line with the published inflation schedule.

**Reply by HarukaMa** [reply]
If it's referring to puzzle difficulty, isn't the retarget algorithm only using `ANCHOR_TIME`, which is 25s? `BLOCK_TIME` seems irrelevant here. 

For coinbase reward, I think we can still completely get rid of `BLOCK_TIME` by only inspecting the timestamp of the block:

```diff
Subject: [PATCH] .
---
Index: ledger/block/src/helpers/target.rs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/ledger/block/src/helpers/target.rs b/ledger/block/src/helpers/target.rs
--- a/ledger/block/src/helpers/target.rs	(revision 02994a1e34e7db8da6c27587537b165aa488868c)
+++ b/ledger/block/src/helpers/target.rs	(date 1730253609794)
@@ -42,16 +41,15 @@
 
 /// Calculates the coinbase reward for a given block.
 ///     R_coinbase = R_anchor(H) * min(P, C_R) / C
-///     R_anchor = Anchor reward at block height.
+///     R_anchor = Anchor reward at timestamp.
 ///     H = Current block height.
 ///     P = Combined proof target.
 ///     C_R = Remaining coinbase target.
 ///     C = Coinbase target.
 pub fn coinbase_reward(
-    block_height: u32,
     starting_supply: u64,
-    anchor_height: u32,
-    block_time: u16,
+    genesis_timestamp: u64,
+    block_timestamp: u64,
     combined_proof_target: u128,
     cumulative_proof_target: u64,
     coinbase_target: u64,
@@ -62,7 +60,7 @@
     let remaining_proof_target = combined_proof_target.min(remaining_coinbase_target as u128);
 
     // Compute the anchor block reward.
-    let anchor_block_reward = anchor_block_reward_at_height(block_height, starting_supply, anchor_height, block_time);
+    let anchor_block_reward = anchor_block_reward_at_timestamp(starting_supply, genesis_timestamp, block_timestamp);
 
     // Calculate the coinbase reward.
     let reward = anchor_block_reward.saturating_mul(remaining_proof_target).saturating_div(coinbase_target as u128);
@@ -75,46 +73,47 @@
     Ok(u64::try_from(reward).expect("Coinbase reward exceeds u64::MAX"))
 }
 
-/// Calculates the anchor block reward for the given block height.
-///     R_anchor = max(floor((2 * S * H_A * H_R) / (H_Y10 * (H_Y10 + 1))), R_Y9).
+/// Calculates the anchor block reward for the given timestamp.
+/// To keep rewards in line with the old algorithm, take old anchor height = 2
+/// and old expected block height at year 10 = 31536000, so the start and end points are unchanged
+///     R_anchor = max(floor((4 * S * T_R) / (T_Y10 * (31536000 + 1))), R_Y9).
 ///     S = Starting supply.
-///     H_A = Anchor block height.
-///     H_R = Remaining number of blocks until year 10.
-///     H_Y10 = Expected block height at year 10.
+///     T_R = Remaining seconds until year 10.
+///     T_Y10 = Total seconds in 10 years.
 ///     R_Y9 = Reward at year 9.
-fn anchor_block_reward_at_height(block_height: u32, starting_supply: u64, anchor_height: u32, block_time: u16) -> u128 {
+fn anchor_block_reward_at_timestamp(starting_supply: u64, genesis_timestamp: u64, block_timestamp: u64) -> u128 {
     // A helper function to calculate the reward at a given block height, without the year 9 baseline.
-    const fn block_reward_at_height(height: u32, starting_supply: u64, anchor_height: u32, block_time: u16) -> u128 {
-        // Calculate the block height at year 10.
-        let block_height_at_year_10 = block_height_at_year(block_time, 10) as u128;
-        // Compute the remaining blocks until year 10.
-        let num_remaining_blocks_to_year_10 = block_height_at_year_10.saturating_sub(height as u128);
+    const fn block_reward_at_timestamp(starting_supply: u64, genesis_timestamp: u64, block_timestamp: u64) -> u128 {
+        // Calculate the number of seconds in a year.
+        const SECONDS_IN_A_YEAR: u128 = 60 * 60 * 24 * 365;
+        // Calculate the timestamp at year 10.
+        let block_timestamp_at_year_10 = timestamp_at_year(genesis_timestamp, 10);
+        // Compute the remaining seconds until year 10.
+        let num_remaining_seconds_to_year_10 = block_timestamp_at_year_10.saturating_sub(block_timestamp);
         // Compute the numerator.
-        let numerator = 2 * starting_supply as u128 * anchor_height as u128 * num_remaining_blocks_to_year_10;
+        let numerator = 4 * starting_supply as u128 * num_remaining_seconds_to_year_10 as u128;
         // Compute the denominator.
-        let denominator = block_height_at_year_10 * (block_height_at_year_10 + 1);
+        let denominator = SECONDS_IN_A_YEAR * 10 * (31536000 + 1);
         // Compute the quotient.
         numerator / denominator
     }
 
     // Calculate the block height at year 9.
-    let block_height_at_year_9 = block_height_at_year(block_time, 9);
+    let block_timestamp_at_year_9 = timestamp_at_year(genesis_timestamp, 9);
     // Compute the unadjusted reward at year 9.
-    let reward_at_year_9 = block_reward_at_height(block_height_at_year_9, starting_supply, anchor_height, block_time);
+    let reward_at_year_9 = block_reward_at_timestamp(starting_supply, genesis_timestamp, block_timestamp_at_year_9);
     // Compute the unadjusted reward at the given block height.
-    let reward_at_block_height = block_reward_at_height(block_height, starting_supply, anchor_height, block_time);
+    let reward_at_block_height = block_reward_at_timestamp(starting_supply, genesis_timestamp, block_timestamp);
     // Compute the anchor block reward.
     reward_at_block_height.max(reward_at_year_9)
 }
 
 /// Returns the block height after a given number of years for a specific block time.
-pub const fn block_height_at_year(block_time: u16, num_years: u32) -> u32 {
+pub const fn timestamp_at_year(genesis_timestamp: u64, num_years: u64) -> u64 {
     // Calculate the number of seconds in a year.
-    const SECONDS_IN_A_YEAR: u32 = 60 * 60 * 24 * 365;
-    // Calculate the one-year block height.
-    let block_height_at_year_1 = SECONDS_IN_A_YEAR / block_time as u32;
-    // Return the block height for the given number of years.
-    block_height_at_year_1 * num_years
+    const SECONDS_IN_A_YEAR: u64 = 60 * 60 * 24 * 365;
+    // Return the timestamp for the given number of years.
+    genesis_timestamp + SECONDS_IN_A_YEAR * num_years
 }
 
 /// Calculate the coinbase target for the given block timestamps and target.
```

This will keep the 10-year total puzzle reward roughly unchanged (2 * starting supply) as long as the retargeting works (not considering the baseline reward output after year 9).

The only remaining usage of `BLOCK_TIME` will be to determine the block numbers in an epoch.

**Reply by howardwu** [reply]
> According to my understanding and what @OzielLa said, if `block_time` will be adjusted to 3 seconds then wouldn't the effect be the same as it is now? It does not reduce the current excessive POS output. At the same time, it will cause the coinbase reward to drop too quickly, which is not in line with the published inflation schedule.

To my knowledge, the puzzle is currently decreasing its emission at 3x the published inflation schedule because the `BLOCK_TIME` is `10` and should be at `3` or `4` to offset the accelerated decay.

**Reply by LoiusZK** [reply]
@howardwu However, the actual block time does not follow 10 seconds, although it is written in the code. You can see from my statistics that the average daily block time is about 3 seconds. It is because the block time is too short that the block height increases too quickly. The PoS output is more than 2 times higher than expected, which leads to the PoW output decreasing much faster than expected, that is, it has reached the level of the 9th year in the 3rd year.

According to the initial design, the fixed PoS reward is 5% of the initial total, that is, 75 million per year. Divided by 365, it is 205,479 per day. Divided by the fixed 23 credits reward per block, the daily block numbers should be 8933. The block time corresponding to this daily block numbers is about 10 seconds. Because according to the 10-second block time, the daily block numbers should be 8640.

So I think the correct approach is to adjust the block time to ensure it is around 10 seconds. 

For PoW, the reward output of each block will inevitably increase because the total number of blocks has decreased, but the total number of PoW rewards per day remains unchanged.
<img width="221" alt="20241101200551" src="https://github.com/user-attachments/assets/b4c7ca5a-d79a-4705-a868-c333ec471b4b">

**Reply by raychu86** [reply]
I believe everyone here is trying to solve the same issue, which is that blocks are currently advancing too quickly and thus reducing the `coinbase_reward` than the initial 10 year plan and emitting more `block_reward` than the declared inflation rate. Let me try to break down the proposed solutions and implications:

The current `coinbase_reward` function is designed to decrease over the course of 10 years and taper off at year 9. This calculation is based on the expected block height at year 10 which is calculated with `BLOCK_TIME = 10`. We will to reach a year 9 steady state at block `NUM_SECONDS_PER_YEAR * NUM_YEARS / BLOCK_TIME = 31,536,000 * 9 / 10 = 28,382,400`. At the current block rate of 3 seconds we will reach that in around 3 years, which is quicker than we intend and gives provers less time overall to find solutions.

## Proposed solutions:

**1. Lower the `BLOCK_TIME` from 10 seconds to 3 seconds.** 

This will align the `coinbase_reward` tapering calculation with the current block rate. The chain should reach year 9 steady state at block `31,536,000 * 9 / 3 = 94,608,000 ` instead of block `28,382,400`. This will align the `coinbase_reward` schedule with the human time scale of 10 years.

This would also reduce the `block_reward` emission per block to closer align with the 5% inflation rate.

The problem with this solution is that if block times deviate from 3 seconds in the future, we have the same problem as before where rewards are not aligned with the timescale and inflation rate: 
- `block_reward` is too large or too small for each block and does not match inflation rate.
- `coinbase_reward` may be decreasing too quickly or too slowly and not aligned with the declared 10 years.

**2. Increase the block time to 10 seconds.**

This solution is in theory possible by enforcing validators to wait X seconds between proposals in an attempt to get the general consensus to approximately 10 seconds. This will slow down blocks enough to closer approximate the correct `block_reward` and `coinbase_reward`.

This is not a great solution because it much more difficult to do in practice because of the variety of factors that affect block production - network traffic, transmission throughput, transmission size, number of validators (which will likely change/increase over time), etc. Not only that, but introducing an artificial wait to validators slows down the capacity of the chain and is definitely more of a bandaid solution than a holistic approach to the problem and to network health.


**3. Use timestamps to dynamically adjust `block_reward` and `coinbase_reward`.** 

The ARC proposal overall is to update `block_reward` to consider the time it took between the previous block and the current block. This will allow us to determine what percentage of the year that block took up and more accurately emit block rewards that align with the 5% inflation rate.

Doing the same with `coinbase_reward` and using `block_timestamp` will allow us to determine where along the 10 year time horizon we are and select the price based on timestamp. This is more accurate than the current block height measurement we had because of the deviation of expected and actual block times. `block_timestamps`, the 10 year `coinbase_reward` scale, and the `block_reward` annual inflation rate all operate on the human timescale, and it makes sense to align them. 


Note that we do not need to change `BLOCK_TIME` here. This is because `BLOCK_TIME` is only has relevance to `block_reward` and `coinbase_reward`, so in theory it can be disregarded for the proposed change if the new reference is timestamps. The target difficulty adjustment algorithm is also unaffected by `BLOCK_TIME` because it only uses `ANCHOR_TIME` (which is set to 25 seconds) to determine changes to `coinbase_target` and `proof_target`.


#### Personal Conclusion:

Overall I believe solution 3 to be the most complete and robust because it seems to adapt to future changes to the network assumptions better. I currently have some draft implementations that I will make public soon for review; there are also a variety of migration considerations to be made.

**Reply by LoiusZK** [reply]
I agree with solution 3, which can ensure that the rewards of both PoS and Coinbase are in accordance with the economic model. But if follow the third solution, will block_time be useless? Is block_time still needed? What is the difference between block_timestamp and block_time? These two seem very similar. Does block_time have anything to do with TPS?

**Reply by HarukaMa** [reply]
> The only remaining usage of `BLOCK_TIME` will be to determine the block numbers in an epoch.

`BLOCK_TIME` is an interval. `block_timestamp` is a timestamp.

And no.

If possible I'd recommend reading the source code to better understand what snarkVM is doing with those variables.

**Reply by LoiusZK** [reply]
Understand. I have another question. Currently, Coinbase rewards are calculated based on block_time and block_height, and it can be ensured that the rewards after the 9th year is based on the block height of the 9th year. So if block_time is replaced with block_timestamp, this should not have much impact, right?

**Reply by LoiusZK** [reply]
According to Haruka's ARC and the solution 3, how to ensure that the total rewards of PoS does not exceed 5% of the initial volume each year? It seems that this is not mentioned in Haruka's ARC and the solution 3. The same problem also exists in coinbase rewards.

**Reply by HarukaMa** [reply]
I strongly believe the original ARC is written to specifically address the 5% issue.

**Reply by raychu86** [reply]
The original ARC proposal is meant to address the 5% inflation rate from `block_rewards` only. In other words, this 5% inflation is with regards to `block_rewards` and not `coinbase_rewards`. 

Solution 3 uses the proposed changes for `block_rewards` to also stabilize `coinbase_rewards` to follow the intentional design. There is not a fixed 5% inflation rate for coinbase rewards. Coinbase rewards are an additional emission that is designed to hit some equilibrium based on the retargetting algorithm and the solutions that make it on chain.

**Reply by raychu86** [reply]
> So if block_time is replaced with block_timestamp, this should not have much impact, right?

As Haruka mentioned - 

`block_time` is the fixed interval that was set in the code (currently set to 10 seconds) that all of the reward calculations are based on. So no matter what the block time was, the reward calculation is based on the 10 seconds.

`block_timestamp` is the actual timestamp of blocks, which gives us granularity and the ability to adjust the rewards based on the actual block time rather than the fixed 10 seconds.

**Reply by LoiusZK** [reply]
> Doing the same with coinbase_reward and using block_timestamp will allow us to determine where along the 10 year time horizon we are and select the price based on timestamp. 

@raychu86 Yes, but as you mentioned in solution 3, the same operation should be performed on coinbase_reward. I understand that it is also necessary to solve the problem that the block height increases too quickly due to the short block time, and the PoW reward is reduced too quickly. That is to say, the reward of the 9th year will be reached by the end of the third year. How do you solve this problem?
![image](https://github.com/user-attachments/assets/0cd579b2-3461-4dcd-b79c-9efc811965e9)

**Reply by raychu86** [reply]
As I mentioned above, **solution 3 already adjusts `coinbase_reward` to be based on timestamp to align the coinbase reward to match the 9 year timeline** and will not be affected by fast or slow block times. 

**To be extra clear:** Coinbase rewards will reach the 9 year mark in 9 years from the genesis block. Even if block times are 1 second or 100 seconds, this will not affect the timeline of coinbase rewards.

**Comment by howardwu** [comment]
For those who are new to this thread and/or following this ARC proposal, I recommend reviewing the technical implementation as proposed by @raychu86 in this PR: https://github.com/AleoNet/snarkVM/pull/2569

It provides a detailed implementation of the transition to the new emission curve as iterated upon from the above ARC discussion threads.
